### PR TITLE
Isolated declarations local inference

### PR DESCRIPTION
**Pull request by dragomirtitian**
Improvements to local inference.

### CODE REVIEW

File: external-declarations/src/compiler/declaration-emit.ts [hunk 1]
```diff
@@ -584,12 +598,19 @@ export function transformDeclarations(context: TransformationContext) {
...
+                try {
+                    enclosingDeclaration = node;
+
+                    const fnTypeNode = factory.createFunctionTypeNode(
+                        visitNodes(fnNode.typeParameters, visitDeclarationSubtree, isTypeParameterDeclaration),
+                        fnNode.parameters.map(p => ensureParameter(p)),
+                        inferReturnType(fnNode).typeNode,
+                    );
+                    return regular(fnTypeNode, node)
+                }finally {
```
Comment (h-joo) [inline external-declarations/src/compiler/declaration-emit.ts]: When can it throw?

File: external-declarations/tests/source/local-inference/return-types-function-declarations.ts [hunk 2]
```diff
@@ -1,83 +1,133 @@
...
-
-export function returnsUnionPrimitive() {
-    if(Math.random() > 0) {
-        return "A";
-    }
-    return "B";
-}
+// @strict: true, false
+// @target: es2022
+// export async function returnsAsyncFunction() {
```
Comment (h-joo) [inline external-declarations/tests/source/local-inference/return-types-function-declarations.ts]: May I ask why these have been commented out?

File: external-declarations/tests/source/local-inference/return-types-function-declarations.ts [hunk 3]
```diff
@@ -1,130 +1,130 @@
 // @strict: true, false
 // @target: es2022
```
Comment (h-joo) [inline external-declarations/tests/source/local-inference/return-types-function-declarations.ts:2]: Oh this has been un-commented now, nice :)

**Review by h-joo — APPROVED** [review]
Thank you for your great work!

Although I haven't fully understood everything from the changes the general look is that this looks good. I see how things are working together, and I don't see any big points that I wanted to address. 

As I was a bit overwhelmed by the size of this PR, my focus of review was less on the mechanical parts like how the emitter was configured to run and such, but on the each logical parts of how it does the inference and so on.

I have some build failures seen from this branch for places that in my opinion it shouldn't have, so I'll send a separate fix for those. 

Again, I really appreciate your work :) thank you!

File: external-declarations/tests/source/local-inference/return-types-function-declarations.ts (hunk 2 above)
Comment (dragomirtitian) [reply external-declarations/tests/source/local-inference/return-types-function-declarations.ts]: Sorry, it was commented by mistake 😅

File: external-declarations/src/compiler/declaration-emit.ts (hunk 1 above)
Comment (dragomirtitian) [reply external-declarations/src/compiler/declaration-emit.ts]: It shouldn't but i used the finally as a way to ensure the `enclosingDeclaration` is always set back to the old value even if there were multiple returns in the body.

**Pull request merged**